onnxscript/rewriter/pattern.py

```diff
@@ -912,23 +876,140 @@ def __init__(self, pattern: GraphPattern) -> None:
         ), "SimplePatternMatcher only supports patterns with a single output node."
         super().__init__(pattern)
 
+    def fail(self, reason: str) -> bool:
+        if self._verbose:
+            if self._matched:  # Print only if at least one node successfully matched.
+                count = len(self._matched)
+                print(f"Match failed after {count} nodes: {reason}")
+        self._match.fail(reason)
+        return False
+
+    def _match_constant(self, pattern_constant: Constant, value: ir.Value) -> bool:
+        """Match a Constant pattern against a value.
+
+        If the constant value is produced by a Constant node, we do not include
+        the constant node as part of the matched graph. Thus, it will not be deleted,
+        if subgraph replacement happens. But subsequent DCE will remove the constant
+        node if it is not used elsewhere.
+        """
+        value = _ir_utils.propagate_const_value(value)
```

Comment:
I would add a comment to remove these two lines when we sort this out

Reply:
Not sure I understand. It serves as documentation, in case a reader starts wondering about this. If we change this design, then we would need to change both the code and documentation.

Reply:
I mean ` value = _ir_utils.propagate_const_value(value)` is temporary and we should remove _ir_utils usage when ready